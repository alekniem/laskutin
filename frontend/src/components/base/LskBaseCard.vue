<template>
    <v-card v-bind="$attrs">
        <slot name="header" />
        <v-divider />
        <slot />
    </v-card>
</template>

<script>
export default {
    name: "LskBaseCard",
};
</script>
