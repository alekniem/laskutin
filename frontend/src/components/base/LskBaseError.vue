<template>
    <div class="red--text font-weight-bold">
        <slot />
    </div>
</template>

<script>
export default {
    name: "LskBaseError",
};
</script>
