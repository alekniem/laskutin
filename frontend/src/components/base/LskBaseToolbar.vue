<template>
    <v-toolbar flat>
        <v-toolbar-title :id="titleId">
            <slot />
        </v-toolbar-title>
        <v-spacer />
        <slot name="button" />
    </v-toolbar>
</template>

<script>
export default {
    name: "LskBaseToolbar",

    props: {
        titleId: { type: String, required: true },
    },
};
</script>
