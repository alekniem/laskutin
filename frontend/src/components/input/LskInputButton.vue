<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
