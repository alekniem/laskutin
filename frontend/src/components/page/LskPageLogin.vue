<template>
    <v-form>
        <v-row>
            <v-col cols="12">
                <LskBaseCard max-width="700">
                    <template #header>
                        <LskBaseToolbar :titleId="cid + 'Title'">
                            {{ "Login" | trans }}
                        </LskBaseToolbar>
                    </template>
                    <v-card-text>
                        <v-row dense>
                            <v-col>
                                <v-text-field
                                    @keyup.enter="focusPasswordField"
                                    :id="cid + 'UsernameField'"
                                    :label="trans('Username')"
                                    ref="usernameField"
                                    v-model="username"
                                />
                            </v-col>
                            <v-col>
                                <v-text-field
                                    @keyup.enter="processLogin"
                                    :id="cid + 'PasswordField'"
                                    :label="trans('Password')"
                                    ref="passwordField"
                                    type="password"
                                    v-model="password"
                                />
                            </v-col>
                        </v-row>
                        <v-row dense>
                            <v-col>
                                <LskInputButton
                                    @click.native="processLogin"
                                    :cid="cid + 'LoginButton'"
                                    buttonText="Login"
                                />
                            </v-col>
                        </v-row>
                        <v-row dense v-if="error">
                            <v-col>
                                <LskBaseError class="pt-4">
                                    {{ error | trans }}
                                </LskBaseError>
                            </v-col>
                        </v-row>
                    </v-card-text>
                </LskBaseCard>
            </v-col>
        </v-row>
    </v-form>
</template>

<script>
import lsk_auth from "@/plugins/lsk_auth";

export default {
    name: "LskPageLogin",

    data() {
        return {
            cid: "loginPage",
            error: "",
            password: "",
            username: "",
        };
    },

    mounted() {
        this.$refs.usernameField.focus();
    },

    methods: {
        focusPasswordField() {
            this.$refs.passwordField.focus();
        },

        async processLogin() {
            await lsk_auth.login(
                this.username,
                this.password,
                (loggedIn, error) => {
                    this.error = error;
                    if (loggedIn) {
                        this.$router.replace({ name: "indexInvoice" });
                    }
                }
            );
        },
    },
};
</script>
