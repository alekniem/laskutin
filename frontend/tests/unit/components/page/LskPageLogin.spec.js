import axios from "axios";
import flushPromises from "flush-promises";
import sinon from "sinon";
import LskBaseCard from "@/components/base/LskBaseCard";
import LskBaseError from "@/components/base/LskBaseError";
import LskBaseToolbar from "@/components/base/LskBaseToolbar";
import LskInputButton from "@/components/input/LskInputButton";
import LskPageLogin from "@/components/page/LskPageLogin";
import { createWrapper } from "UT/lsk_test_helper";

jest.mock("axios");

axios.get.mockImplementation((url) => {
    if (url === "sanctum/csrf-cookie") {
        return Promise.resolve({
            data: {
                message: "CSRF COOKIE QUERY OK",
            },
        });
    }
    return Promise.reject({
        response: {
            data: {
                message: "GET QUERY ERROR",
            },
        },
    });
});

axios.post.mockImplementation((url, data) => {
    if (url === "login" && data.username === "test" && data.password === "valid") {
        return Promise.resolve({
            data: {
                message: "LOGIN QUERY OK",
            },
        });
    }
    return Promise.reject({
        response: {
            data: {
                message: "POST QUERY ERROR",
            },
        },
    });
});

const stubs = {
    LskBaseCard,
    LskBaseError,
    LskBaseToolbar,
    LskInputButton,
};

describe("LskPageLogin", () => {
    beforeEach(() => {
        if (localStorage.getItem("token")) {
            localStorage.removeItem("token");
        }
    });

    afterEach(() => {
        sinon.restore();
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskPageLogin, {
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("should handle valid login", async () => {
        const wrapper = createWrapper(LskPageLogin, {
            stubs,
        });
        await flushPromises();
        wrapper.vm.$router = { replace: sinon.spy() };
        wrapper.find("#loginPageUsernameField").setValue("test");
        wrapper.find("#loginPagePasswordField").setValue("valid");
        await flushPromises();
        wrapper.find("#loginPageLoginButton").trigger("click");
        await flushPromises();
        const param = {
            name: "indexInvoice",
        };
        expect(wrapper.vm.$router.replace.calledWith(param)).toBe(true);
        expect(wrapper.vm._data.error).toBe("");
    });

    it("should handle invalid login", async () => {
        const wrapper = createWrapper(LskPageLogin, {
            stubs,
        });
        await flushPromises();
        wrapper.vm.$router = { replace: sinon.spy() };
        wrapper.find("#loginPageUsernameField").setValue("test");
        wrapper.find("#loginPagePasswordField").setValue("invalid");
        await flushPromises();
        wrapper.find("#loginPageLoginButton").trigger("click");
        await flushPromises();
        expect(wrapper.vm.$router.replace.called).toBe(false);
        expect(wrapper.vm._data.error).toBe("Failed to login");
    });
});
