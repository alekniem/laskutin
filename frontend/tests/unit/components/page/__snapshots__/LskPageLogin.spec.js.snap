// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskPageLogin should match snapshot 1`] = `
<form novalidate="novalidate"
      class="v-form"
>
  <div class="row">
    <div class="col col-12">
      <div class="v-card v-sheet theme--light"
           style="max-width: 700px;"
           max-width="700"
      >
        <header class="v-sheet theme--light v-toolbar v-toolbar--flat"
                style="height: 64px;"
        >
          <div class="v-toolbar__content"
               style="height: 64px;"
          >
            <div id="loginPageTitle"
                 class="v-toolbar__title"
            >
              Login
            </div>
            <div class="spacer">
            </div>
          </div>
        </header>
        <hr role="separator"
            aria-orientation="horizontal"
            class="v-divider theme--light"
        >
        <div class="v-card__text">
          <div class="row row--dense">
            <div class="col">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="loginPageUsernameField"
                             class="v-label theme--light"
                             style="left: 0px; position: absolute;"
                      >
                        Username
                      </label>
                      <input id="loginPageUsernameField"
                             type="text"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="col">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="loginPagePasswordField"
                             class="v-label theme--light"
                             style="left: 0px; position: absolute;"
                      >
                        Password
                      </label>
                      <input id="loginPagePasswordField"
                             type="password"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="row row--dense">
            <div class="col">
              <button type="button"
                      class="v-btn v-btn--has-bg theme--light v-size--small primary"
                      id="loginPageLoginButton"
              >
                <span class="v-btn__content">
                  Login
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
